### ISSUE TITLE
Statements must be separated by newlines or semicolons - many occurrences

### ISSUE THREAD

**Issue by SGP57** [microsoft/pylance-release#1521]
<h3> Environment data </h3>

-   Language Server version: 2021.6.3
-   OS and version: linux x64
-   Python version (and distribution if applicable, e.g. Anaconda): 
-   python.analysis.indexing: undefined
-   python.analysis.typeCheckingMode: off

<h3> Expected behaviour </h3>

Correct analysis!

<h3> Actual behaviour </h3>

Many false positives.

I am seeing many error reports, mostly this one but others too. It's reporting this error in comments at the start of the file as well.

Python language is selected.

I have <N> for a new file, selected python and pasted in my 39 lines of program and errors are not reported. I put an test-error into the new file and that was reported successfully, and removed when I removed my test-error .

So, it's not the actual syntax etc. I'm relatively new at this. What assumptions are you making that I wouldn't have predicted? I guess there's all sorts of things about my setup that I could discover and add but this seems a bit fundamental so I'd appreciate suggestions / guidance.

python --version
Python 3.8.10

lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 20.04.2 LTS
Release:	20.04
Codename:	focal

<h3> Logs </h3>

Python Language Server Log
```
XXX 'Please paste the output from your clipboard'
```

<h3> Code Snippet / Additional information </h3>

```python
XXX
```

**Comment by erictraut** [comment]
We can't diagnose a problem without seeing your code. Please include a minimal, self-contained code sample that exhibits the problem.

One common cause of the "statements must be separated by newlines" is the use of Python 2 code. Pylance supports only Python 3.x. If you are using Python 2 code, you should choose a different language server like Jedi.

**Comment by SGP57** [comment]
Re python version: it's 3.8.10

Re giving sample code that exhibits the problem: that was the whole point of the fault report. The code in the original file did show the false errors but copying and pasting the same code into a new file/tab in vscode did not show the errors. Consequently, showing the false errors was unlikely to happen in code that I submitted.

Within my original file, I have now deleted everything (Ctrl+A, Ctrl+X) and then reinserted (Ctrl+V) and the errors have gone. I'm guessing therefore your diagnosis will be 'not a fault'. There clearly was an issue here - I'm guessing that pylance got a bit confused and thought that you may be interested in that. 

I do see that a screenshot would probably have been useful.

The 'Statements must be separated by newlines or semicolons' messages gave locations: sometimes several locations (word boundaries) on the same line in # comment lines. Also, sometimes the location was a few columns beyond the end of the line and there wasn't any white space present there, especially in import statements. There was also a message about an invalid character of (I think it was) '\u21' but nothing unexpected was at that location.

**Comment by bschnurr** [comment]
Just to re-iterate,

We will need a screenshot or code sample to try to reproduce what you're seeing.

**Comment by SGP57** [comment]
Apologies. Not possible.

As I fixed the problem by deleting and pasting the same code, it probably wasn't just that code. I've tried <Ctrl+Z> many times to undo to see if I could recreate but the errors are not showing up, so unfortunately a screenshot isn't possible either now.

I get that it was going to be hard to reproduce. I had thought it might be useful for you to get this report in case you (plural) had seen or do see anything similar.

I understand you will probably close this.

**Comment by SGP57** [comment]
Maybe the real point was that the errors that I saw didn't make sense. Several "must be separated" errors about the same comment line and errors that gave locations that didn't exist - just a bit over the end of the line.

**Comment by erictraut** [comment]
I can think of two possible causes:
1. VS Code opened the file with the wrong encoding (e.g. UTF16 vs UTF8)
2. There were bad, unprintable characters within the file — perhaps it was corrupted on disk

In either case, I don't think this is a bug in pylance.

**Comment by SGP57** [comment]
I see your thinking. I'd agree with your conclusion.

This was a new "program". I had written some python but had also copied and pasted from web examples and python doc and edited. I'd suggest that the file on disk was ok because the python program ran successfully but the pasting from the web might open up possibilities about character encodings. An interesting restriction within vscode if that is the case! However, the cut and paste presumably sorted that out.

Thanks.

**Comment by jbeyer16** [comment]
Not sure that it's too helpful, but sorta related along the lines of "errors that don't make sense". I do at times run into phantom lints, or where it appears pylance thinks text exists in a cell that doesn't. The only way to clear it out is to paste the code into a new cell. It's hard to consistently reproduce, it just sometimes gets into this weird state and i end up reloading the notebook to clear it.

Here are a view images. In this case it is linting comments. 

![image](https://user-images.githubusercontent.com/19380579/125142119-d31fed80-e0cb-11eb-9059-5a3b16a6fd5b.png)

![image](https://user-images.githubusercontent.com/19380579/125142129-da46fb80-e0cb-11eb-83e7-dfaaa26a2f93.png)

Here it's linting blank space (px is imported and the cell runs fine).

![image](https://user-images.githubusercontent.com/19380579/125142220-14180200-e0cc-11eb-9b58-11e60882cfcd.png)

Here it's alerting to something not related to the line of code.

![image](https://user-images.githubusercontent.com/19380579/125142294-3dd12900-e0cc-11eb-9f39-408efe05076f.png)

**Comment by jakebailey** [comment]
Have you been able to consistently reproduce it, or is it seemingly random? If you can reproduce it, I'd be pretty in the LSP trace (`"python.trace.server": "verbose"`) to see what the edits are.

**Comment by jbeyer16** [comment]
Once it gets in that state the cells remain that way and as i edit them different phantom lints appear, but i haven't yet been able to figure out what gets it there.

**Comment by jakebailey** [comment]
Yeah, it's unfortunately the trace up to that point that's the most interesting. I'd think this would be in the vein of a bug in the Jupyter extension, except that the original issue was not in a notebook.

**Comment by jbeyer16** [comment]
Gotcha. Can you clarify this?

> LSP trace ("python.trace.server": "verbose")

Is that not something to put in settings.json? I'm happy to pay more attention to it if it happens again.

**Comment by jakebailey** [comment]
Yes, it would be; it's undocumented so VS Code will squiggle it, but it's working. You'll see a load more output in the Python Language Server output panel.

This may significantly slow down your editor, as it logs big JSON objects for every single thing between the editor and Pylance, but it's worth a try if you can get it to the bad state.

**Comment by jbeyer16** [comment]
ah okay gotcha. Yeah was confused since it said it was invalid. Okay I'll keep in on there and see if I'm able to catch anything! Thanks!

**Comment by jbeyer16** [comment]
Hmm, in theory I've triggered it again.

Is this helpful at all? It seems as though it's "lost" a line in a cell. Hovering over the first line `import n1805` triggers the range saying line 0 and thinking it's on line 1 where `import numpy as np` is?

![image](https://user-images.githubusercontent.com/19380579/125144503-f5693980-e0d2-11eb-8ada-640e5ce19425.png)

**Comment by jbeyer16** [comment]
Able to see the same error that OP mentioned too.

![image](https://user-images.githubusercontent.com/19380579/125144841-08c8d480-e0d4-11eb-8392-8ae1abdb1503.png)

**Comment by jakebailey** [comment]
It's the full log I'm interested in, as it would contain every file change sent from the client to us so we can see if it's consistent or if our incremental text handling is somehow off (in Pylance or in Jupyter).

**Comment by whimian** [comment]
I got a lot of warnings for my code after updating to the new version of pylance. So I tried a basic class definition, The following is the code and the warnings I got from Pylance. 

<img width="695" alt="Snipaste_2021-07-13_11-05-52" src="https://user-images.githubusercontent.com/3821278/125384139-08c61e80-e3cb-11eb-8e8c-71a37ab988ed.png">

**Comment by Sti2nd** [comment]
Here are more examples of Pylance linting wrong:

The code:
![image](https://user-images.githubusercontent.com/6373430/126777576-7f152225-b7e9-4963-aedd-60cd212d478a.png)
The lint errors:
![image](https://user-images.githubusercontent.com/6373430/126777606-2617e0d8-5cfc-4fc1-a1e9-ebc1784f0377.png)
![image](https://user-images.githubusercontent.com/6373430/126777518-11229d22-fb46-44f9-8382-8e7bd1c21d9e.png)

**Comment by jbeyer16** [comment]
@jakebailey The attached log is from startup to when i noticed an issue present. It's pretty contrived, but it seems to be along the lines of what i see at times in my standard workflow.

My first thought in regards to the conditions needed to generate it is maybe something related to the plotly figures containing so much data that the language server falls behind? Not sure if others that are experiencing it use plotly, but it seems somewhat reproducible for me to have a bunch of large plotly figures showing inline then start editing code to have an error, fix that error, but see the pylance lint as if the error is still there.

Let me know if there's more I can do!

![image](https://user-images.githubusercontent.com/19380579/128574732-26236890-a94f-4f42-8cfe-42d0a283827d.png)
![image](https://user-images.githubusercontent.com/19380579/128574752-57935d24-3206-47a6-9647-17fbfa44d1fb.png)

[log2.txt](https://github.com/microsoft/pylance-release/files/6948106/log2.txt)

**Comment by jakebailey** [comment]
In your case, do these messages go away eventually? I recall that plotly is one of those libraries with an incredible number of modules with lots of complicated code, which makes it hard to analyze, so using plotly may be the problem itself and things lag behind as we try to keep up.

We don't know anything about the figures themselves, only the document that the Jupyter extension produces that looks like the notebook concatenated into a single file.

**Comment by jbeyer16** [comment]
They don't seem to go away, but they do seem to migrate. I can delete a cell that has them but the phantom error messages will just move on to the cell that bumps up from the deletion.

I don't know too much about plotly internals, but i do know their figures are essentially JSON objects. So i was thinking that if the representation of the notebook that is sent to the language server contains cell output, it'd have a massive additional JSON object to look through. If you're not looking at cell output though i guess that's irrelevant.

But, for example, if i save an `ipynb` file and look at it in a text editor, all those data points are in there as a JSON object.

**Comment by jakebailey** [comment]
We don't work with the cell outputs or the ipynb; we get simulated Python files that contain only Python code. The figures themselves are a red herring, and this would have to be some sort of desync we need to look at the log to try and figure out.

**Comment by erictraut** [comment]
Thanks for the trace logs. Those were very helpful. I spent some time walking through each message to reconstruct the document one edit at a time.

The "textDocument/publishDiagnostics" input messages look correct to me. They're in the right order, and when I apply the deltas, I reconstruct the text that you show in your screen shot above.

The "textDocument/publishDiagnostics" output messages also look correct to me. They include the diagnostics I would expect to see while you're coding — various syntax errors due to incomplete code. The text ranges for these diagnostics also correspond to the line and column numbers that I would expect from the reconstructed contents.

Subsequent "textDocument/publishDiagnostics" output messages should replace the previously-reported list of diagnostics within the document. That doesn't seem to be happening. For example, the "textDocument/publishDiagnostics" message sent at timestamp 2:52:12 contains the message `Expected ":"`. The subsequent "textDocument/publishDiagnostics" message at timestamp 2:52:13 no longer contains a diagnostic with this message because the intervening edits have eliminated that syntax error condition. The client is supposed to replace older diagnostic lists with an updated list when it receives a newer "textDocument/publishDiagnostics" message. 

Based on the evidence, it looks like the problem is in the code that provides support for jupyter notebooks — the code that combines the cells in the notebook and writes the combined Python code to a file for pylance to analyze. I'm guessing that this code is not properly handling the "textDocument/publishDiagnostics" messages that it receives. It should process them in the order they're received, and newer diagnostic lists should replace older lists.

@jakebaily, one thing that I noticed is that pylance is not including a `version` in the "textDocument/publishDiagnostics" message. The LSP protocol indicates that such a version is not required, but I wonder if the absence of this version is triggering a bug in the jupyter notebook client code.

**Comment by jakebailey** [comment]
It's very possible that the version would help; it's been something we've wanted to plumb correctly as an improvement.

Given the way that these work, it's very possible that it could be as simple as there being an async step somewhere between the translation of the faked documents back to the cells, such that every now and then, they get scheduled out of order. Versioning would fix that, so long as the translation makes use of them properly (and that, I don't know).

**Comment by imRajAryan09** [comment]
Pylance is checking and false flagging syntax errors for the text written in markdown and comments. I am using VS Code for writing a Jupyter Notebook

**Comment by jakebailey** [comment]
@jbeyer16 Does your notebook happen to have non-Python cells in it? E.g. markdown cells?

**Comment by jbeyer16** [comment]
@jakebailey Yep it did!

**Comment by jakebailey** [comment]
Aha, then I think that what's happening to everyone here is that when a non-Python cell is edited, the code that handles converting the Python parts of the notebook get confused. It could be that it's sending an invalid change, or is expecting a version number to increase (even though from our POV there were no edits), or is sending the wrong edits due to cell confusion, etc. Good to have a lead.

**Comment by jbeyer16** [comment]
oh that's an interesting insight! I think most of my notebooks typically have Markdown, which explains why i see it a good bit. Glad you've got a place to start now!

**Comment by roboserg** [comment]
Started a new, empty notebook. Didnt use any markdown cells or jupyter magic commands, still Pylance is confused - https://i.imgur.com/M7aDYNt.png

**Comment by jakebailey** [comment]
@roboserg Your issue appears unrelated; the diagnostics are not misplaced or incorrect. This typically means that the kernel you've selected for your notebook is not the same as the interpreter you've selected at the bottom left for the Python extension.

**Comment by roboserg** [comment]
> @roboserg Your issue appears unrelated; the diagnostics are not misplaced or incorrect. This typically means that the kernel you've selected for your notebook is not the same as the interpreter you've selected at the bottom left for the Python extension.

Yes, those were my thoughts. I triple checked to have the correct conda env (I even manually re-selected it). Nothing helped till I restarted VS Code. I'd assume re-selecting the proper python conda env would load it, but it did'nt.
